cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/CreateOperation.java

```diff
@@ -0,0 +1,174 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.strimzi.controller.cluster.operations;
+
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.extensions.Deployment;
+import io.fabric8.kubernetes.api.model.extensions.StatefulSet;
+import io.fabric8.openshift.api.model.BuildConfig;
+import io.fabric8.openshift.api.model.ImageStream;
+import io.strimzi.controller.cluster.K8SUtils;
+import io.strimzi.controller.cluster.OpenShiftUtils;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class CreateOperation<U, R extends HasMetadata> implements Operation<U> {
+
+    private static final Logger log = LoggerFactory.getLogger(CreateOperation.class);
+
+    private final R resource;
+    private final String resourceKind;
+
+
+
+    public CreateOperation(String resourceKind, R resource) {
+        this.resourceKind = resourceKind;
+        this.resource = resource;
+    }
+
+    @Override
+    public void execute(Vertx vertx, U utils, Handler<AsyncResult<Void>> handler) {
+        vertx.createSharedWorkerExecutor("kubernetes-ops-pool").executeBlocking(
+                future -> {
+                    if (!exists(utils, resource.getMetadata().getNamespace(), resource.getMetadata().getName())) {
+                        try {
+                            log.info("Creating {} {}", resourceKind, resource);
+                            create(utils, resource);
+                            future.complete();
+                        } catch (Exception e) {
+                            log.error("Caught exception while creating {}", resourceKind, e);
+                            future.fail(e);
+                        }
+                    }
+                    else {
+                        log.warn("{} {} already exists", resourceKind, resource);
+                        future.complete();
+                    }
+                },
+                false,
+                res -> {
+                    if (res.succeeded()) {
+                        log.info("{} {} has been created", resourceKind, resource);
+                        handler.handle(Future.succeededFuture());
+                    }
+                    else {
+                        log.error("{} creation failed: {}", resourceKind, res.cause().toString());
+                        handler.handle(Future.failedFuture(res.cause()));
+                    }
+                }
+        );
+    }
+
+    protected abstract void create(U utils, R resource);
```

Comment:
One of the things which we discussed in the past with @ppatierno is that we need to make this more synchronous (at least for some operations). Would you put that into the `create` method? Or would you build that as a separate operation? (e.g. you would have two operations ... one for creating StatefulSet and one for making sure the Statefulset is running)

Also, do you think we still need the `K8SUtils` and `OpenShiftUtils` classes as an additional layer of abstraction between the operations and Fabric8 API?

Reply:
TBH I've not really understood the code the point of having an opinion, yet. I just saw that all the create operations were doing the same basic thing and couldn't resist getting rid of the duplicated code. I suppose that if the `*Utils` classes are only ever used from the Operation classes, then the `*Utils` really isn't very useful, and we should factor them out.

For the creation operation specifically, I would be inclined to avoid the check-and-create pattern, and switch to create-and-catch the exception if it already exists (assuming we can distinguish that error condition from others). On a hunch I suspect having a generic 'Is running' operation would be most flexible in terms of API. 

But I will look at it in more detail today.